Fix flakey bart robot unit test

One of our tests for the bart robot occasionally fails. We should fix this

`tests/devices/unit_tests/test_bart_robot.py::test_given_program_not_running_and_pin_unmounts_then_mounts_when_load_pin_then_pin_loaded - 
AssertionError: Expected 'trigger' to have been called once. Called 0 times.`

Or see https://github.com/DiamondLightSource/dodal/actions/runs/12165707719/job/33930319285


## Acceptance Criteria
* Test made more robust

--- a/src/dodal/devices/robot.py
+++ b/src/dodal/devices/robot.py
@@ -21,6 +21,9 @@ from ophyd_async.epics.core import (
 
 from dodal.log import LOGGER
 
+WAIT_FOR_OLD_PIN_MSG = "Waiting on old pin unloaded"
+WAIT_FOR_NEW_PIN_MSG = "Waiting on new pin loaded"
+
 
 class RobotLoadFailed(Exception):
     error_code: int
@@ -144,6 +147,7 @@ class BartRobot(StandardReadable, Movable[SampleLocation]):
             # in the current task, when it propagates to here we should cancel all pending tasks before bubbling up
             for task in tasks:
                 task.cancel()
+
             raise
 
     async def _load_pin_and_puck(self, sample_location: SampleLocation):
@@ -164,9 +168,9 @@ class BartRobot(StandardReadable, Movable[SampleLocation]):
         )
         await self.load.trigger()
         if await self.gonio_pin_sensor.get_value() == PinMounted.PIN_MOUNTED:
-            LOGGER.info("Waiting on old pin unloaded")
+            LOGGER.info(WAIT_FOR_OLD_PIN_MSG)
             await wait_for_value(self.gonio_pin_sensor, PinMounted.NO_PIN_MOUNTED, None)
-        LOGGER.info("Waiting on new pin loaded")
+        LOGGER.info(WAIT_FOR_NEW_PIN_MSG)
 
         await self.pin_mounted_or_no_pin_found()
 

--- a/tests/devices/unit_tests/test_bart_robot.py
+++ b/tests/devices/unit_tests/test_bart_robot.py
@@ -1,5 +1,6 @@
 import traceback
-from asyncio import create_task, sleep
+from asyncio import create_task
+from functools import partial
 from unittest.mock import ANY, AsyncMock, MagicMock, call, patch
 
 import pytest
@@ -16,8 +17,8 @@ from dodal.devices.robot import BartRobot, PinMounted, RobotLoadFailed, SampleLo
 
 async def _get_bart_robot() -> BartRobot:
     device = BartRobot("robot", "-MO-ROBOT-01:")
-    device.LOAD_TIMEOUT = 0.05  # type: ignore
-    device.NOT_BUSY_TIMEOUT = 0.05  # type: ignore
+    device.LOAD_TIMEOUT = 1  # type: ignore
+    device.NOT_BUSY_TIMEOUT = 1  # type: ignore
     await device.connect(mock=True)
     return device
 
@@ -29,11 +30,10 @@ async def test_bart_robot_can_be_connected_in_sim_mode():
 
 async def test_given_robot_load_times_out_when_load_called_then_exception_contains_error_info():
     device = await _get_bart_robot()
+    device.LOAD_TIMEOUT = 0.01  # type: ignore
+    device.NOT_BUSY_TIMEOUT = 0.01  # type: ignore
 
-    async def _sleep(*args):
-        await sleep(1)
-
-    device._load_pin_and_puck = AsyncMock(side_effect=_sleep)
+    device._load_pin_and_puck = AsyncMock(side_effect=TimeoutError)
 
     set_mock_value(device.prog_error.code, (expected_error_code := 10))
     set_mock_value(device.prog_error.str, (expected_error_string := "BAD"))
@@ -50,6 +50,8 @@ async def test_given_program_running_when_load_pin_then_logs_the_program_name_an
     patch_logger: MagicMock,
 ):
     device = await _get_bart_robot()
+    device.LOAD_TIMEOUT = 0.01  # type: ignore
+    device.NOT_BUSY_TIMEOUT = 0.01  # type: ignore
     program_name = "BAD_PROGRAM"
     set_mock_value(device.program_running, True)
     set_mock_value(device.program_name, program_name)
@@ -94,24 +96,34 @@ async def test_given_program_not_running_and_pin_unmounting_but_new_pin_not_moun
         raise
 
 
-async def set_with_happy_path(device: BartRobot) -> AsyncStatus:
+def _set_pin_sensor_on_log_messages(device: BartRobot, msg):
+    if msg == "Waiting on old pin unloaded":
+        set_mock_value(device.gonio_pin_sensor, PinMounted.NO_PIN_MOUNTED)
+    elif msg == "Waiting on new pin loaded":
+        set_mock_value(device.gonio_pin_sensor, PinMounted.PIN_MOUNTED)
+
+
+# Use log info messages to determine when to set the gonio_pin_sensor, so we don't have to use any sleeps during testing
+async def set_with_happy_path(
+    device: BartRobot, mock_log_info: MagicMock
+) -> AsyncStatus:
     """Mocks the logic that the robot would do on a successful load"""
-    device.LOAD_TIMEOUT = 0.05  # type: ignore
+
+    mock_log_info.side_effect = partial(_set_pin_sensor_on_log_messages, device)
+    device.LOAD_TIMEOUT = 1  # type: ignore
+    device.NOT_BUSY_TIMEOUT = 1  # type: ignore
     set_mock_value(device.program_running, False)
     set_mock_value(device.gonio_pin_sensor, PinMounted.PIN_MOUNTED)
     status = device.set(SampleLocation(15, 10))
-    await sleep(0.025)
-
-    set_mock_value(device.gonio_pin_sensor, PinMounted.NO_PIN_MOUNTED)
-    await sleep(0.025)
-
-    set_mock_value(device.gonio_pin_sensor, PinMounted.PIN_MOUNTED)
     return status
 
 
-async def test_given_program_not_running_and_pin_unmounts_then_mounts_when_load_pin_then_pin_loaded():
+@patch("dodal.devices.robot.LOGGER.info")
+async def test_given_program_not_running_and_pin_unmounts_then_mounts_when_load_pin_then_pin_loaded(
+    mock_log_info: MagicMock,
+):
     device = await _get_bart_robot()
-    status = await set_with_happy_path(device)
+    status = await set_with_happy_path(device, mock_log_info)
     await status
     assert status.success
     assert (await device.next_puck.get_value()) == 15
@@ -121,10 +133,8 @@ async def test_given_program_not_running_and_pin_unmounts_then_mounts_when_load_
 
 async def test_given_waiting_for_pin_to_mount_when_no_pin_mounted_then_error_raised():
     device = await _get_bart_robot()
-    status = create_task(device.pin_mounted_or_no_pin_found())
-    await sleep(0.2)
     set_mock_value(device.prog_error.code, 25)
-    await sleep(0.01)
+    status = create_task(device.pin_mounted_or_no_pin_found())
     with pytest.raises(RobotLoadFailed):
         await status
 
@@ -132,18 +142,18 @@ async def test_given_waiting_for_pin_to_mount_when_no_pin_mounted_then_error_rai
 async def test_given_waiting_for_pin_to_mount_when_pin_mounted_then_no_error_raised():
     device = await _get_bart_robot()
     status = create_task(device.pin_mounted_or_no_pin_found())
-    await sleep(0.01)
     set_mock_value(device.gonio_pin_sensor, PinMounted.PIN_MOUNTED)
-    await sleep(0.01)
     await status
 
 
 @patch("dodal.devices.robot.wait_for")
 async def test_set_waits_for_both_timeouts(mock_wait_for: AsyncMock):
     device = await _get_bart_robot()
-    device._load_pin_and_puck = MagicMock()
+    device.LOAD_TIMEOUT = 0.01  # type: ignore
+    device.NOT_BUSY_TIMEOUT = 0.01  # type: ignore
+    device._load_pin_and_puck = MagicMock()  # type: ignore
     await device.set(SampleLocation(1, 2))
-    mock_wait_for.assert_awaited_once_with(ANY, timeout=0.1)
+    mock_wait_for.assert_awaited_once_with(ANY, timeout=0.02)
 
 
 async def test_moving_the_robot_will_reset_error_if_light_curtain_is_tripped_and_still_throw_if_error_not_cleared():
@@ -164,7 +174,10 @@ async def test_moving_the_robot_will_reset_error_if_light_curtain_is_tripped_and
     )
 
 
-async def test_moving_the_robot_will_reset_error_if_light_curtain_is_tripped_and_continue_if_error_cleared():
+@patch("dodal.devices.robot.LOGGER.info")
+async def test_moving_the_robot_will_reset_error_if_light_curtain_is_tripped_and_continue_if_error_cleared(
+    mock_log_info: MagicMock,
+):
     device = await _get_bart_robot()
     set_mock_value(device.controller_error.code, BartRobot.LIGHT_CURTAIN_TRIPPED)
 
@@ -173,7 +186,7 @@ async def test_moving_the_robot_will_reset_error_if_light_curtain_is_tripped_and
         lambda *_, **__: set_mock_value(device.controller_error.code, 0),
     )
 
-    await (await set_with_happy_path(device))
+    await (await set_with_happy_path(device, mock_log_info))
 
     get_mock(device).assert_has_calls(
         [

--- a/tests/devices/unit_tests/test_bart_robot.py
+++ b/tests/devices/unit_tests/test_bart_robot.py
@@ -66,6 +66,8 @@ async def test_given_program_not_running_but_pin_not_unmounting_when_load_pin_th
     patch_logger: MagicMock,
 ):
     device = await _get_bart_robot()
+    device.LOAD_TIMEOUT = 0.01  # type: ignore
+    device.NOT_BUSY_TIMEOUT = 0.01  # type: ignore
     set_mock_value(device.program_running, False)
     set_mock_value(device.gonio_pin_sensor, PinMounted.PIN_MOUNTED)
     device.load = AsyncMock(side_effect=device.load)
@@ -81,6 +83,8 @@ async def test_given_program_not_running_and_pin_unmounting_but_new_pin_not_moun
     patch_logger: MagicMock,
 ):
     device = await _get_bart_robot()
+    device.LOAD_TIMEOUT = 0.01  # type: ignore
+    device.NOT_BUSY_TIMEOUT = 0.01  # type: ignore
     set_mock_value(device.program_running, False)
     set_mock_value(device.gonio_pin_sensor, PinMounted.NO_PIN_MOUNTED)
     device.load = AsyncMock(side_effect=device.load)
@@ -158,6 +162,8 @@ async def test_set_waits_for_both_timeouts(mock_wait_for: AsyncMock):
 
 async def test_moving_the_robot_will_reset_error_if_light_curtain_is_tripped_and_still_throw_if_error_not_cleared():
     device = await _get_bart_robot()
+    device.LOAD_TIMEOUT = 0.01  # type: ignore
+    device.NOT_BUSY_TIMEOUT = 0.01  # type: ignore
     set_mock_value(device.controller_error.code, BartRobot.LIGHT_CURTAIN_TRIPPED)
 
     with pytest.raises(RobotLoadFailed) as e:

--- a/tests/devices/unit_tests/test_bart_robot.py
+++ b/tests/devices/unit_tests/test_bart_robot.py
@@ -12,7 +12,14 @@ from ophyd_async.testing import (
     set_mock_value,
 )
 
-from dodal.devices.robot import BartRobot, PinMounted, RobotLoadFailed, SampleLocation
+from dodal.devices.robot import (
+    WAIT_FOR_NEW_PIN_MSG,
+    WAIT_FOR_OLD_PIN_MSG,
+    BartRobot,
+    PinMounted,
+    RobotLoadFailed,
+    SampleLocation,
+)
 
 
 async def _get_bart_robot() -> BartRobot:
@@ -23,6 +30,12 @@ async def _get_bart_robot() -> BartRobot:
     return device
 
 
+# For tests which are intentionally triggering a timeout error
+def _set_fast_robot_timeouts(robot: BartRobot):
+    robot.LOAD_TIMEOUT = 0.01  # type: ignore
+    robot.NOT_BUSY_TIMEOUT = 0.01  # type: ignore
+
+
 async def test_bart_robot_can_be_connected_in_sim_mode():
     device = await _get_bart_robot()
     await device.connect(mock=True)
@@ -30,9 +43,7 @@ async def test_bart_robot_can_be_connected_in_sim_mode():
 
 async def test_given_robot_load_times_out_when_load_called_then_exception_contains_error_info():
     device = await _get_bart_robot()
-    device.LOAD_TIMEOUT = 0.01  # type: ignore
-    device.NOT_BUSY_TIMEOUT = 0.01  # type: ignore
-
+    _set_fast_robot_timeouts(device)
     device._load_pin_and_puck = AsyncMock(side_effect=TimeoutError)
 
     set_mock_value(device.prog_error.code, (expected_error_code := 10))
@@ -50,8 +61,7 @@ async def test_given_program_running_when_load_pin_then_logs_the_program_name_an
     patch_logger: MagicMock,
 ):
     device = await _get_bart_robot()
-    device.LOAD_TIMEOUT = 0.01  # type: ignore
-    device.NOT_BUSY_TIMEOUT = 0.01  # type: ignore
+    _set_fast_robot_timeouts(device)
     program_name = "BAD_PROGRAM"
     set_mock_value(device.program_running, True)
     set_mock_value(device.program_name, program_name)
@@ -66,8 +76,7 @@ async def test_given_program_not_running_but_pin_not_unmounting_when_load_pin_th
     patch_logger: MagicMock,
 ):
     device = await _get_bart_robot()
-    device.LOAD_TIMEOUT = 0.01  # type: ignore
-    device.NOT_BUSY_TIMEOUT = 0.01  # type: ignore
+    _set_fast_robot_timeouts(device)
     set_mock_value(device.program_running, False)
     set_mock_value(device.gonio_pin_sensor, PinMounted.PIN_MOUNTED)
     device.load = AsyncMock(side_effect=device.load)
@@ -83,8 +92,7 @@ async def test_given_program_not_running_and_pin_unmounting_but_new_pin_not_moun
     patch_logger: MagicMock,
 ):
     device = await _get_bart_robot()
-    device.LOAD_TIMEOUT = 0.01  # type: ignore
-    device.NOT_BUSY_TIMEOUT = 0.01  # type: ignore
+    _set_fast_robot_timeouts(device)
     set_mock_value(device.program_running, False)
     set_mock_value(device.gonio_pin_sensor, PinMounted.NO_PIN_MOUNTED)
     device.load = AsyncMock(side_effect=device.load)
@@ -101,9 +109,9 @@ async def test_given_program_not_running_and_pin_unmounting_but_new_pin_not_moun
 
 
 def _set_pin_sensor_on_log_messages(device: BartRobot, msg: str):
-    if msg == "Waiting on old pin unloaded":
+    if msg == WAIT_FOR_OLD_PIN_MSG:
         set_mock_value(device.gonio_pin_sensor, PinMounted.NO_PIN_MOUNTED)
-    elif msg == "Waiting on new pin loaded":
+    elif msg == WAIT_FOR_NEW_PIN_MSG:
         set_mock_value(device.gonio_pin_sensor, PinMounted.PIN_MOUNTED)
 
 
@@ -114,8 +122,6 @@ async def set_with_happy_path(
     """Mocks the logic that the robot would do on a successful load"""
 
     mock_log_info.side_effect = partial(_set_pin_sensor_on_log_messages, device)
-    device.LOAD_TIMEOUT = 1  # type: ignore
-    device.NOT_BUSY_TIMEOUT = 1  # type: ignore
     set_mock_value(device.program_running, False)
     set_mock_value(device.gonio_pin_sensor, PinMounted.PIN_MOUNTED)
     status = device.set(SampleLocation(15, 10))
@@ -138,7 +144,7 @@ async def test_given_program_not_running_and_pin_unmounts_then_mounts_when_load_
 async def test_given_waiting_for_pin_to_mount_when_no_pin_mounted_then_error_raised():
     device = await _get_bart_robot()
     set_mock_value(device.prog_error.code, 25)
-    status = create_task(device.pin_mounted_or_no_pin_found())
+    status = device.pin_mounted_or_no_pin_found()
     with pytest.raises(RobotLoadFailed):
         await status
 
@@ -153,8 +159,7 @@ async def test_given_waiting_for_pin_to_mount_when_pin_mounted_then_no_error_rai
 @patch("dodal.devices.robot.wait_for")
 async def test_set_waits_for_both_timeouts(mock_wait_for: AsyncMock):
     device = await _get_bart_robot()
-    device.LOAD_TIMEOUT = 0.01  # type: ignore
-    device.NOT_BUSY_TIMEOUT = 0.01  # type: ignore
+    _set_fast_robot_timeouts(device)
     device._load_pin_and_puck = MagicMock()  # type: ignore
     await device.set(SampleLocation(1, 2))
     mock_wait_for.assert_awaited_once_with(ANY, timeout=0.02)
@@ -162,8 +167,7 @@ async def test_set_waits_for_both_timeouts(mock_wait_for: AsyncMock):
 
 async def test_moving_the_robot_will_reset_error_if_light_curtain_is_tripped_and_still_throw_if_error_not_cleared():
     device = await _get_bart_robot()
-    device.LOAD_TIMEOUT = 0.01  # type: ignore
-    device.NOT_BUSY_TIMEOUT = 0.01  # type: ignore
+    _set_fast_robot_timeouts(device)
     set_mock_value(device.controller_error.code, BartRobot.LIGHT_CURTAIN_TRIPPED)
 
     with pytest.raises(RobotLoadFailed) as e:

--- a/tests/devices/unit_tests/test_bart_robot.py
+++ b/tests/devices/unit_tests/test_bart_robot.py
@@ -100,7 +100,7 @@ async def test_given_program_not_running_and_pin_unmounting_but_new_pin_not_moun
         raise
 
 
-def _set_pin_sensor_on_log_messages(device: BartRobot, msg):
+def _set_pin_sensor_on_log_messages(device: BartRobot, msg: str):
     if msg == "Waiting on old pin unloaded":
         set_mock_value(device.gonio_pin_sensor, PinMounted.NO_PIN_MOUNTED)
     elif msg == "Waiting on new pin loaded":

